### PR TITLE
Add support for MrDeepfakes

### LINKED ISSUE #23685: [Site-Support-Request] MrDeepFakes.com (NSFW)
## Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running youtube-dl version **2020.01.01**
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs violate any copyrights
- [X] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Single video: https://mrdeepfakes.com/video/5939/salma-hayek-lesbian-short-clip

## Description

The generic extractor unfortunately fails to extract the video.

### LINKED ISSUE #30764: Add support for MrDeepfakes
## Checklist

- [ x] I'm reporting a new site support request
- [ x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x ] I've checked that all provided URLs are alive and playable in a browser
- [ x] I've checked that none of provided URLs violate any copyrights
- [ x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs
https://mrdeepfakes.com/video/20/jessica-alba-dildo-masturbate-deep-fakes

## Description
I tried the extractor from this PR https://github.com/ytdl-org/youtube-dl/pull/22390
but i got error:
Traceback (most recent call last):
File "c:\python\python39\lib\runpy.py", line 197, in _run_module_as_main
return run_code(code, main_globals, None,
File "c:\python\python39\lib\runpy.py", line 87, in run_code
exec(code, run_globals)
File "C:\Python\Python39\Scripts\youtube-dl.exe_main.py", line 7, in
File "c:\python\python39\lib\site-packages\youtube_dl_init.py", line 475, in main
real_main(argv)
File "c:\python\python39\lib\site-packages\youtube_dl_init.py", line 465, in _real_main
retcode = ydl.download(all_urls)
File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 2068, in download
res = self.extract_info(
File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 808, in extract_info
return self.__extract_info(url, ie, download, extra_info, process)
File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 815, in wrapper
return func(self, *args, **kwargs)
File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 836, in __extract_info
ie_result = ie.extract(url)
File "c:\python\python39\lib\site-packages\youtube_dl\extractor\common.py", line 534, in extract
ie_result = self._real_extract(url)
File "c:\python\python39\lib\site-packages\youtube_dl\extractor\mrdeepfakes.py", line 114, in _real_extract
license_code = self._license_code_decrypt(license_code)
File "c:\python\python39\lib\site-packages\youtube_dl\extractor\mrdeepfakes.py", line 59, in _license_code_decrypt
k = int(f[:j + 1])
TypeError: slice indices must be integers or None or have an index method

### PR DESCRIPTION
**Pull request by ghost**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added support for extracting public videos from MrDeepfakes.

Resolves #23685.
Resolves #30764.

### CODE REVIEW

**Comment by afterdelight** [comment]
Got error:
----------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python\python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\python\python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Python\Python39\Scripts\youtube-dl.exe\__main__.py", line 7, in <module>
  File "c:\python\python39\lib\site-packages\youtube_dl\__init__.py", line 475, in main
    _real_main(argv)
  File "c:\python\python39\lib\site-packages\youtube_dl\__init__.py", line 465, in _real_main
    retcode = ydl.download(all_urls)
  File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 2068, in download
    res = self.extract_info(
  File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 808, in extract_info
    return self.__extract_info(url, ie, download, extra_info, process)
  File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "c:\python\python39\lib\site-packages\youtube_dl\YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "c:\python\python39\lib\site-packages\youtube_dl\extractor\common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "c:\python\python39\lib\site-packages\youtube_dl\extractor\mrdeepfakes.py", line 114, in _real_extract
    license_code = self._license_code_decrypt(license_code)
  File "c:\python\python39\lib\site-packages\youtube_dl\extractor\mrdeepfakes.py", line 59, in _license_code_decrypt
    k = int(f[:j + 1])
TypeError: slice indices must be integers or None or have an __index__ method
----------------------------------------------------------------------------------------------------------------------------------
Can Anyone fix?

**Comment by 89z** [comment]
https://github.com/ytdl-org/youtube-dl/issues/30764

File: youtube_dl/extractor/mrdeepfakes.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import ExtractorError
```
Comment (dirkf) [inline youtube_dl/extractor/mrdeepfakes.py:5]:
```suggestion
import itertools
import re

from .common import InfoExtractor
from ..utils import (
    determine_ext,
    ExtractorError,
    parse_resolution,
    url_or_none,
)

**Review by dirkf — CHANGES_REQUESTED** [review]
Taking over from Ghost

**Comment by afterdelight** [comment]
does that changes support downloading from favorites, playlist or watch later?

**Comment by 89z** [comment]
@dirkf I wouldn't waste anymore time on this. Sadly I think I picked the wrong hill to die on, as @afterdelight is quite obviously a help vampire [1], and ungrateful to the point of it being confusing. However if you want to continue, of course that's your decision.

1. https://meta.stackoverflow.com/a/252077

**Comment by afterdelight** [comment]
no updates?

**Review by afterdelight — APPROVED** [review]

**Comment by afterdelight** [comment]
got error with your suggestion change
`SyntaxError: closing parenthesis '}' does not match opening parenthesis '('`

from this line
`flashvars = self._search_regex(r'(?s)\bvar\s+flashvars\s*=\s*\({'.+?})\s*(?:;|<script)', webpage, 'flashvars')`

**Comment by dirkf** [comment]
Looks like a typo: 
```diff
-flashvars = self._search_regex(r'(?s)\bvar\s+flashvars\s*=\s*\({'.+?})\s*(?:;|<script)', webpage, 'flashvars')
+flashvars = self._search_regex(r'(?s)\bvar\s+flashvars\s*=\s*\({.+?})\s*(?:;|<script)', webpage, 'flashvars')
```

**Comment by afterdelight** [comment]
now I get after removing `'`:
```
[MrDeepfakes] 10088: Downloading webpage
Traceback (most recent call last):
  File "C:\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Python\Python310\Scripts\youtube-dl.exe\__main__.py", line 7, in <module>
  File "C:\Python\Python310\lib\site-packages\youtube_dl\__init__.py", line 475, in main
    _real_main(argv)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\__init__.py", line 465, in _real_main
    retcode = ydl.download(all_urls)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\YoutubeDL.py", line 2080, in download
    res = self.extract_info(
  File "C:\Python\Python310\lib\site-packages\youtube_dl\YoutubeDL.py", line 809, in extract_info
    return self.__extract_info(url, ie, download, extra_info, process)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\YoutubeDL.py", line 816, in wrapper
    return func(self, *args, **kwargs)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\YoutubeDL.py", line 837, in __extract_info
    ie_result = ie.extract(url)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\extractor\common.py", line 535, in extract
    ie_result = self._real_extract(url)
  File "C:\Python\Python310\lib\site-packages\youtube_dl\extractor\mrdeepfakes.py", line 105, in _real_extract
    flashvars = self._search_regex(r'(?s)\bvar\s+flashvars\s*=\s*\({.+?})\s*(?:;|<script)', webpage, 'flashvars')
  File "C:\Python\Python310\lib\site-packages\youtube_dl\extractor\common.py", line 992, in _search_regex
    mobj = re.search(pattern, string, flags)
  File "C:\Python\Python310\lib\re.py", line 200, in search
    return _compile(pattern, flags).search(string)
  File "C:\Python\Python310\lib\re.py", line 303, in _compile
    p = sre_compile.compile(pattern, flags)
  File "C:\Python\Python310\lib\sre_compile.py", line 764, in compile
    p = sre_parse.parse(p, flags)
  File "C:\Python\Python310\lib\sre_parse.py", line 964, in parse
    raise source.error("unbalanced parenthesis")
re.error: unbalanced parenthesis at position 35
```

**Comment by afterdelight** [comment]
mrdeepfakes.py codes:
```
# coding: utf-8
from __future__ import unicode_literals

import itertools
import re
from .common import InfoExtractor
from ..utils import (
    determine_ext,
    ExtractorError,
    parse_resolution,
    url_or_none,
)


class MrDeepfakesIE(InfoExtractor):
    _VALID_URL = r'https?://(?:www\.)?mrdeepfakes\.com/video/(?P<id>[0-9]+)/.+'
    _TEST = {
        'url': 'https://mrdeepfakes.com/video/5/selena-gomez-pov-deep-fakes',
        'md5': 'fec4ad5ec150f655e0c74c696a4a2ff4',
        'info_dict': {
            'id': '5',
            'ext': 'mp4',
            'title': 'Selena Gomez POV (Deep Fakes)',
            'description': 'Deepfake Selena Gomez lets you fuck her tight shaved pussy',
            'height': 720,
            'age_limit': 18
        }
    }

    

    def _flashvars_get_format(self, flashvars, key, fatal=True):
        """Returns tuple of format (url, quality) for format with URL key 'key', or (None, None) if it doesn't exist."""
        url = flashvars.get(key)
        if url is None:
            if fatal:
                raise ExtractorError('Unable to extract flashvars ' + key)
            return (None, None)
        url = url.split('/', 2)[-1]
        quality = parse_resolution(flashvars.get(key + '_text') or url).get('height')
        return (url, quality)

    def _license_code_decrypt(self, license_code, license_code_length=16):
        """Decrypts license code from form in flashvars to form usable by _uuid_decrypt."""
        if license_code != '':
            f = ''

            for char in license_code[1:]:
                if char == '0':
                    f += '1'
                else:
                    f += char

            j = len(f) // 2
            k = int(f[:j + 1])
            l = int(f[j:])
            g = abs(l - k)
            f = g
            g = abs(k - l)
            f += g
            f *= 2
            f = str(f)
            i = license_code_length // 2 + 2
            m = ''

            for g in range(j + 1):
                for h in range(1, 5):
                    n = int(license_code[g + h]) + int(f[g])
                    if n >= i:
                        n -= i
                    m += str(n)
            return m

        return license_code

    def _uuid_decrypt(self, uuid, license_code):
        """Decrypts the uuid section of a file URL, using a decrypted license code."""
        h = uuid[:32]

        j = h

        for k in range(len(h) - 1, -1, -1):
            l = k
            for m in range(k, len(license_code)):
                l += int(license_code[m])
            while(l >= len(h)):
                l -= len(h)
            n = ''
            for o in range(len(h)):
                if o == k:
                    n += str(h[l])
                else:
                    if o == l:
                        n += str(h[k])
                    else:
                        n += str(h[o])
            h = n
        uuid = uuid.replace(j, h)
        return uuid

    def _real_extract(self, url):
        video_id = self._match_id(url)
        webpage = self._download_webpage(url, video_id)
        title = self._og_search_title(webpage)
        flashvars = self._search_regex(r'(?s)\bvar\s+flashvars\s*=\s*\({.+?})\s*(?:;|<script)', webpage, 'flashvars')
        flashvars = self._parse_json(flashvars, video_id) or {}
        license_code = flashvars.get('license_code')

        if license_code is None:
            # Support for accounts is as yet absent
            self.raise_login_required()

        license_code = self._license_code_decrypt(license_code)

        encrypted_formats = []
        encrypted_formats.append(self._flashvars_get_format(flashvars, 'video_url'))

        for ix in itertools.count(1):
            key = 'video_alt_url'
            if ix > 1:
                key += '%d' % (ix, )
            format = self._flashvars_get_format(flashvars, key, fatal=False)

            if None not in format:
                encrypted_formats.append(format)
            else:
                break

        decrypted_formats = []

        for url, quality in encrypted_formats:
            url_split = url.split('/')

            for ix, chunk in enumerate(url_split):
                if len(chunk) == 42:
                    chosen = ix
                    break
            else:
                continue

            url_split[chosen] = self._uuid_decrypt(url_split[chosen], license_code)
            decrypted_url = '/'.join(url_split)
            decrypted_formats.append((decrypted_url, quality))

        # Sort by quality
        decrypted_formats.sort(key=lambda x: x[1])

        formats = []

        for url, quality in decrypted_formats:
            formats.append({
                'url': url,
                'height': quality,
                'ext': determine_ext(url, 'mp4'),
                'http_headers': {
                    'Referer': 'https://mrdeepfakes.com/'
                }
            })
        
        self._sort_formats(formats)
        
        return {
            'id': video_id,
            'title': title,
            'description': self._og_search_description(webpage),
            'thumbnail': url_or_none(flashvars.get('preview_url')) or self._og_search_thumbnail(webpage),
            'categories': filter(None, re.split(r'\s*,\s*', flashvars.get('video_categories', ''))),
            'tags': filter(None, re.split(r'\s*,\s*', flashvars.get('video_tags', ''))),
            'age_limit': 18,
            'formats': formats,
        }
```

**Comment by dirkf** [comment]
Try the new git master.

**Comment by afterdelight** [comment]
it worked for single link. what about user channel?

**Comment by dirkf** [comment]
Currently, we have to create a separate extractor to handle playlist pages, as in the ThisVid extractor, which could be used as an example.

Or an interested person could extract the URLs from the page in some other way (__grep__, __findstr__, eg) and make up a batch file that can be passed to yt-dl.

**Comment by afterdelight** [comment]
yes, that will do

**Pull request closed without merging**